canary/canary_test.py

```diff
@@ -0,0 +1,75 @@
+import asyncio
+import datetime
+import os
+import time
+
+from pymongo import MongoClient
+from typing import Coroutine, Any
+from viam.robot.client import RobotClient
+from viam.components.board import Board
+
+async def connect(robot_address: str, api_key: str, api_key_id: str) -> Coroutine[Any, Any, RobotClient]:
+    opts = RobotClient.Options.with_api_key(
+      api_key=api_key,
+      api_key_id=api_key_id
+    )
+    return await RobotClient.at_address(robot_address, opts)
+
+async def main():
+    mongo_connection_str = os.environ["MONGODB_TEST_OUTPUT_URI"]
+    db_client = MongoClient(mongo_connection_str)
+    db = db_client["esp32_canary"]
+    coll = db["hourly_results"]
+
+    timestamp = datetime.datetime.now()
+
+    default_item = {
+        "_id": timestamp,
+        "connection_success": False,
+        "board_api_success": False,
+        "error": "",
+        "connection_latency_ms": 0
+    }
+
+    robot_address = os.environ["ESP32_CANARY_ROBOT"]
+    api_key = os.environ["ESP32_CANARY_API_KEY"]
+    api_key_id = os.environ["ESP32_CANARY_API_KEY_ID"]
+
+    print(f"connecting to robot at {robot_address} ...")
+
+    start = time.time()
+
+    for i in range(5):
+        try:
+            robot = await connect(robot_address, api_key, api_key_id)
+            break
+        except Exception as e:
+            if i == 4:
+                default_item["error"] = str(e)
+                coll.insert_one(default_item)
+                raise e
+            print(e)
+        time.sleep(0.5)
+
+    connectivity_time = (time.time() - start) * 1000
+    default_item["connection_success"] = True
+    default_item["connection_latency_ms"] = connectivity_time
```

Comment:
This will end up reporting the total connection time over multiple connection attempts, if such is necessary. Nothing wrong with that, but it could be misleading / noisier than it needs to be. Maybe report both? The total time required to connect across retries, and the latency of the succeeding connection?

Reply:
Does it functionally matter what the latency of the last attempt is versus the overall time to someone interacting via the SDK who has to try multiple times anyway? Would saving the number of attempts instead be useful?

Reply:
Functionally, no, but I think we might be interested to know if we suddenly went from a situation where it mostly connects on first attempt and takes 100ms, to a situation where it typically takes 4 failing tries at 10ms each, and then connects taking 60ms. As things stand we wouldn't see that transition because the average wouldn't change. Reporting the number of attempts required to connect seems like a useful way to disambiguate those cases, and if that's easier to implement then sure, I'm happy with that approach too. I also don't feel strongly about it, so I'm fine moving forward with just the average if you prefer.

Reply:
I agree with @acmorrow, this metric can be clearer -- connection_latency_ms - latency of the successful connection, if available.